atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/store/AtlasDbTransactionStore.java

```diff
@@ -0,0 +1,181 @@
+        return transactionManager.runTaskWithRetry(task -> {
+            Map<Cell, byte[]> values = task.get(tableReference, Set.of(atlasCell));
+            return Optional.ofNullable(values.get(atlasCell)).map(Ints::fromByteArray);
+        });
+    }
+
+    @Override
+    public Optional<WitnessedTransaction> readWrite(List<TransactionAction> actions) {
+        AtomicReference<Long> startTimestampReference = new AtomicReference<>();
+        AtomicReference<List<WitnessedTransactionAction>> witnessedActionsReference = new AtomicReference<>();
+        try {
+            transactionManager.runTaskWithRetry(txn -> {
+                AtlasDbTransactionActionVisitor visitor = new AtlasDbTransactionActionVisitor(txn);
+                startTimestampReference.set(txn.getTimestamp());
+                witnessedActionsReference.set(actions.stream()
+                        .sequential()
+                        .map(action -> action.accept(visitor))
+                        .collect(Collectors.toList()));
+                return null;
+            });
+
+            TransactionStatus status =
+                    transactionManager.getTransactionService().getV2(startTimestampReference.get());
+
+            Optional<Long> commitTimestamp = TransactionStatuses.getCommitTimestamp(status);
+
+            return Optional.of(ImmutableWitnessedTransaction.builder()
+                    .startTimestamp(startTimestampReference.get())
+                    .commitTimestamp(commitTimestamp)
+                    .actions(witnessedActionsReference.get())
+                    .build());
+        } catch (Exception e) {
+            log.info("Failed to record transaction due to an exception", e);
+            return Optional.empty();
+        }
+    }
+
+    private class AtlasDbTransactionActionVisitor implements TransactionActionVisitor<WitnessedTransactionAction> {
+
+        private final Transaction transaction;
+
+        public AtlasDbTransactionActionVisitor(Transaction transaction) {
+            this.transaction = transaction;
+        }
+
+        @Override
+        public WitnessedTransactionAction visit(ReadTransactionAction readTransactionAction) {
+            Cell cell = toAtlasCell(readTransactionAction.cell());
+            Map<Cell, byte[]> cells =
+                    transaction.get(tableReference, Set.of(toAtlasCell(readTransactionAction.cell())));
+            Optional<Integer> value = Optional.ofNullable(cells.get(cell)).map(Ints::fromByteArray);
+            return readTransactionAction.witness(value);
+        }
+
+        @Override
+        public WitnessedTransactionAction visit(WriteTransactionAction writeTransactionAction) {
+            transaction.put(
+                    tableReference,
+                    Map.of(
+                            toAtlasCell(writeTransactionAction.cell()),
+                            Ints.toByteArray(writeTransactionAction.value())));
+            transaction.put(
```

Comment:
This is 🍝 . I think long-term we can better address this by providing some list of table(s) we desire to our TransactionStore, and have our actions reference a `table`. It is then up to the workflow to create/manage indexes, or some other abstraction, which is a bit cleaner. That said, I think this is OK for now, considering we will only have one schema-type available.

Reply:
ok this has been dealt with, workflow needs to deal with indexes, not the responsibility of the underlying store.